File: app/src/main/java/com/marinov/colegioetapa/UpdateChecker.java
package com.marinov.colegioetapa;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.stream.Collectors;

public class UpdateChecker {
    private static final String TAG = "UpdateChecker";
    private static final String PREFS_NAME = "UpdatePrefs";
    private static final String KEY_LAST_VERSION = "last_version";

    public interface UpdateListener {
        void onUpdateAvailable(String url);
        void onUpToDate();
        void onError(String message);
    }

    public static void checkForUpdate(Context context, UpdateListener listener) {
        new Thread(() -> {
            try {
                URL url = new URL("https://api.github.com/repos/gmb7886/EtapaClient/releases/latest");
                HttpURLConnection conn = (HttpURLConnection) url.openConnection();
                conn.setRequestMethod("GET");
                conn.setRequestProperty("Accept", "application/vnd.github.v3+json");
                conn.setRequestProperty("User-Agent", "EtapaClient-Android");
                conn.setConnectTimeout(10000);

                if (conn.getResponseCode() == HttpURLConnection.HTTP_OK) {
                    String json = readResponseStream(conn);
                    JSONObject release = new JSONObject(json);
                    String latestVersion = release.getString("tag_name");
                    String currentVersion = BuildConfig.VERSION_NAME;

                    SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
                    String lastNotifiedVersion = prefs.getString(KEY_LAST_VERSION, "");

                    if (!latestVersion.equals(currentVersion) && !latestVersion.equals(lastNotifiedVersion)) {
                        prefs.edit().putString(KEY_LAST_VERSION, latestVersion).apply();
                        listener.onUpdateAvailable(release.getString("html_url"));
                    } else {
                        listener.onUpToDate();
                    }
                } else {
                    listener.onError("HTTP error: " + conn.getResponseCode());
                }
            } catch (Exception e) {
                Log.e(TAG, "Erro na verificação", e);
                listener.onError(e.getMessage());
            }
        }).start();
    }

    private static String readResponseStream(HttpURLConnection conn) throws Exception {
        try (InputStream is = conn.getInputStream();
             BufferedReader br = new BufferedReader(new InputStreamReader(is))) {
            return br.lines().collect(Collectors.joining("\n"));
        }
    }
}

File: app/src/main/java/com/marinov/colegioetapa/UpdateChecker.kt
package com.marinov.colegioetapa

import android.content.Context
import android.util.Log
import androidx.core.content.edit
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.util.stream.Collectors

object UpdateChecker {
    private const val TAG = "UpdateChecker"
    private const val PREFS_NAME = "UpdatePrefs"
    private const val KEY_LAST_VERSION = "last_version"

    interface UpdateListener {
        fun onUpdateAvailable(url: String)
        fun onUpToDate()
        fun onError(message: String)
    }

    fun checkForUpdate(context: Context, listener: UpdateListener) {
        Thread {
            runCatching {
                val url = URL("https://api.github.com/repos/gmb7886/EtapaClient/releases/latest")
                val conn = url.openConnection() as HttpURLConnection
                conn.apply {
                    requestMethod = "GET"
                    setRequestProperty("Accept", "application/vnd.github.v3+json")
                    setRequestProperty("User-Agent", "EtapaClient-Android")
                    connectTimeout = 10000
                }

                when (conn.responseCode) {
                    HttpURLConnection.HTTP_OK -> {
                        val json = readResponseStream(conn)
                        val release = JSONObject(json)
                        val latestVersion = release.getString("tag_name")
                        val currentVersion = BuildConfig.VERSION_NAME

                        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                        val lastNotifiedVersion = prefs.getString(KEY_LAST_VERSION, "") ?: ""

                        if (isVersionGreater(latestVersion, currentVersion)) {
                            if (latestVersion != lastNotifiedVersion) {
                                prefs.edit { putString(KEY_LAST_VERSION, latestVersion) }
                                listener.onUpdateAvailable(release.getString("html_url"))
                            } else {
                                listener.onUpToDate()
                            }
                        } else {
                            listener.onUpToDate()
                        }
                    }
                    else -> listener.onError("HTTP error: ${conn.responseCode}")
                }
            }.onFailure { e ->
                Log.e(TAG, "Erro na verificação", e)
                listener.onError(e.message ?: "Erro desconhecido")
            }
        }.start()
    }

    private fun readResponseStream(conn: HttpURLConnection): String {
        return conn.inputStream.use { inputStream ->
            BufferedReader(InputStreamReader(inputStream)).use { reader ->
                reader.lines().collect(Collectors.joining("\n"))
            }
        }
    }

    private fun normalizeVersion(version: String): String {
        return version.replace(Regex("^[^0-9]+"), "")
    }

    fun isVersionGreater(newVersion: String, currentVersion: String): Boolean {
        val normalizedNew = normalizeVersion(newVersion)
        val normalizedCurrent = normalizeVersion(currentVersion)

        val newParts = normalizedNew.split(".").map { it.toIntOrNull() ?: 0 }
        val currentParts = normalizedCurrent.split(".").map { it.toIntOrNull() ?: 0 }

        for (i in 0 until maxOf(newParts.size, currentParts.size)) {
            val newPart = newParts.getOrElse(i) { 0 }
            val currentPart = currentParts.getOrElse(i) { 0 }

            when {
                newPart > currentPart -> return true
                newPart < currentPart -> return false
            }
        }
        return false
    }
}